File: Sample/Koudai/Server/src/ZyGames.Tianjiexing.BLL/Action/Action1444.cs
/****************************************************************************
Copyright (c) 2013-2015 scutgame.com

http://www.scutgame.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
using System.Collections.Generic;
using ZyGames.Framework.Common;
using ZyGames.Framework.Game.Service;
using ZyGames.Tianjiexing.Model;
using ZyGames.Tianjiexing.Model.Config;
using ZyGames.Framework.Game.Cache;
using ZyGames.Tianjiexing.Model.Enum;
using System;

namespace ZyGames.Tianjiexing.BLL.Action
{

    /// <summary>
    /// 1444_佣兵升级卡片列表接口
    /// </summary>
    public class Action1444 : BaseAction
    {
        private int pageIndex;
        private int pageSize;
        private int pageCount;
        private int itemNum;
        private List<UserItemInfo> useritemList = new List<UserItemInfo>();

        public Action1444(ZyGames.Framework.Game.Contract.HttpGet httpGet)
            : base(ActionIDDefine.Cst_Action1444, httpGet)
        {

        }

        public override void BuildPacket()
        {
            this.PushIntoStack(pageCount);
            this.PushIntoStack(useritemList.Count);
            foreach (var item in useritemList)
            {
                ItemBaseInfo itemInfo = new ConfigCacheSet<ItemBaseInfo>().FindKey(item.ItemID);
                itemNum = UserItemHelper.CheckItemNum(ContextUser.UserID, item.ItemID);
                DataStruct dsItem = new DataStruct();
                dsItem.PushIntoStack(item.UserItemID.ToNotNullString());
                dsItem.PushIntoStack(item.ItemID);
                dsItem.PushIntoStack(itemInfo == null ? string.Empty : itemInfo.ItemName.ToNotNullString());
                dsItem.PushIntoStack(itemInfo == null ? string.Empty : itemInfo.HeadID.ToNotNullString());
                dsItem.PushIntoStack(itemNum);
                dsItem.PushIntoStack(itemInfo == null ? 0 : itemInfo.EffectNum);
                
                this.PushIntoStack(dsItem);
            }

        }

        public override bool GetUrlElement()
        {
            if (httpGet.GetInt("PageIndex", ref pageIndex)
                 && httpGet.GetInt("PageSize", ref pageSize))
            {
                return true;
            }
            return false;
        }

        public override bool TakeAction()
        {
            var package = UserItemPackage.Get(ContextUser.UserID);
            if (package != null)
            {
                var useritemArray = package.ItemPackage.FindAll(s => s.ItemType == ItemType.DaoJu && s.PropType == 14);
                useritemList = useritemArray.GetPaging(pageIndex, pageSize, out pageCount);
            }
            return true;
        }
    }
}

File: Sample/Koudai/Server/release/Script/PyScript/Action/action1444.py
import clr, sys
clr.AddReference('ZyGames.Framework.Common');
clr.AddReference('ZyGames.Framework');
clr.AddReference('ZyGames.Framework.Game');
clr.AddReference('ZyGames.Tianjiexing.Model');
clr.AddReference('ZyGames.Tianjiexing.BLL');
clr.AddReference('ZyGames.Tianjiexing.Lang');
clr.AddReference('ZyGames.Tianjiexing.Component');
from action import *
from System import *
from System.Collections.Generic import *
from ZyGames.Framework.Common.Log import *
from ZyGames.Tianjiexing.Model.ConfigModel import *
from ZyGames.Framework.Common import *
from ZyGames.Tianjiexing.Model import *
from ZyGames.Tianjiexing.BLL import *
from ZyGames.Tianjiexing.Lang import *
from ZyGames.Framework.Game.Cache import *
from ZyGames.Framework.Game.Service import *
from ZyGames.Framework.Game.Runtime import *
from ZyGames.Framework.Common import *
from ZyGames.Framework.Cache.Generic import *
from ZyGames.Tianjiexing.Model.Config import *
from ZyGames.Tianjiexing.Model.Enum import *
from ZyGames.Tianjiexing.BLL.Base import *
from ZyGames.Tianjiexing.Component import *
from ZyGames.Framework.Common.Serialization import *

# 1444_佣兵升级卡片列表接口
class UrlParam:
    Result = True;
    pageIndex = 0;
    pageSize = 0;   
    itemNum = 0;
    userID = ''; 

class ActionResult(DataResult):
    def __init__(self):
        DataResult.__init__(self)
        self.pageCount = 1
        self.useritemList = []
        self.userItemNum=1      #使用经验卡的数量   

def getUrlElement(httpGet, parent):
    urlParam = UrlParam();
    if httpGet.Contains("PageIndex")\
    and httpGet.Contains("PageSize"):
        urlParam.pageIndex = httpGet.GetIntValue("PageIndex", 1, 100 );
        urlParam.pageSize = httpGet.GetIntValue("PageSize");
    else:
        urlParam.Result = False;
    return urlParam;

def takeAction(urlParam, parent):
    actionResult = ActionResult();
    userId = parent.Current.User.PersonalId;
    urlParam.userID = parent.Current.User.PersonalId;
    contextUser = parent.Current.User;
    package = UserItemPackage.Get(userId);
    if package != None:
        useritemArray = package.ItemPackage.FindAll(lambda s : s.ItemType == ItemType.DaoJu and s.PropType == 14);   
        result = MathUtils.GetPaging[UserItemInfo](useritemArray,urlParam.pageIndex,urlParam.pageSize);
        if result :
           actionResult.useritemList =  result[0];
           actionResult.pageCount = result[1];
    actionResult.userItemNum=1                  #界面使用经验卡数量默认为1
    return actionResult;

def buildPacket(writer, urlParam, actionResult):
    itemName ='';
    headId='';
    effectNum =0;
    writer.PushIntoStack(actionResult.pageCount);
    writer.PushIntoStack(actionResult.useritemList.Count);
    for item in actionResult.useritemList:
        itemInfo = ConfigCacheSet[ItemBaseInfo]().FindKey(item.ItemID);
        #itemNum = UserItemHelper.CheckItemNum(urlParam.userID, item.ItemID);
        if itemInfo != None:
            itemName = itemInfo.ItemName;
            headId = itemInfo.HeadID;
            effectNum = itemInfo.EffectNum;
        dsItem = DataStruct();
        dsItem.PushIntoStack(MathUtils.ToNotNullString(item.UserItemID));
        dsItem.PushIntoStack(item.ItemID);
        dsItem.PushIntoStack(MathUtils.ToNotNullString(itemName));
        dsItem.PushIntoStack(MathUtils.ToNotNullString(headId));
        dsItem.PushIntoStack(item.Num);
        dsItem.PushIntoStack(effectNum);
        dsItem.PushIntoStack(actionResult.userItemNum);     #界面使用经验卡数量默认为1
        writer.PushIntoStack(dsItem);
    return True;